File: aoc2311.ts
let day = 11
import { config } from "https://deno.land/x/dotenv/mod.ts"
config()
let session = Deno.env.get("AOC_SESSION")

let get_stuff = (day): Promise<string> => {

  let url: string = `https://adventofcode.com/2023/day/${day}/input`
  console.log(session)
  let headers = new Headers({ 'Cookie': `session=${session}` })
  return fetch(url, { headers })
        .then((resp) => resp.text())
        .catch((err) => { throw err })
}

get_stuff(day).then((infile) => {

  let lines: string[] = infile.split('\n')
  lines.pop()
  let part1 = 0, part2 = 0

  console.log(lines[0])

  // Solver

  let coor: number[][] = []
  let R = lines.length, C = lines[0].length, r = -1, c 
  while (++r < R) {
    c = -1
    while (++c < C) {
      let node = lines[r][c]
      if (node == '#')
        coor.push([r, c])
    }
  }
  let EC:number[] = []
  let ER:number[] = []
  let IA:string[][] = []
  c = -1
  while (++c < C) {
    r = -1
    let temp: string[] = []
    while (++r < R) {
      temp.push(lines[r][c])
    }
    IA.push(temp)
  }
  r = -1
  while (++r < lines.length) {
    let ok = true
    for (let node of lines[r])
      if (node != '.')
        ok = false
    if (ok)
      ER.push(r)
  }
  c = -1
  while (++c < IA.length) {
    let ok = true
    for (let node of IA[c])
      if (node != '.')
        ok = false
    if (ok)
      EC.push(c)
  }

  const Calc = (coor: number[][], ER:number[], EC:number[], xp: number = 2): number => {

    // if (coor.length == 0) return 0
    let res = 0, i = -1
    while (++i < coor.length) {

      let [r, c] = coor[i]
      for (let node_next of coor.slice(0, i)) {

        let [next_r, next_c] = node_next
        let [s, e] = [ Math.min(r, next_r), Math.max(r, next_r) ]
        let node = s - 1
        while (++node < e) {
          if (ER.includes(node))
            res += xp - 1
        }
        res += e - s
        s = Math.min(c, next_c), e = Math.max(c, next_c)
        node = s - 1
        while (++node < e) {
          if (EC.includes(node))
            res += xp - 1
        }
        res += e - s
      }
    }
    return res
  }

  part1 = Calc (coor, ER, EC)
  part2 = Calc (coor, ER, EC, 1e6)
  console.log("part 1:", part1)
  console.log("part 2:", part2)

}).catch((err) => { throw err })

const checkTwoArraysAreEqual = (L:number[], R:number[]): boolean =>
{ return L.length === R.length && L.every((val, idx) => val === R[idx]) }

const checkSetHas = (set: Set<number[]>, this_array: number[]): boolean =>
{
  for (let that_array of set)
    if (checkTwoArraysAreEqual(this_array, that_array))
      return true
  return false
}

File: aoc2311.py
F = 0
r1, r2 = 0, 0
A = []
nul = '.'
with open('11.' + str(F)) as file:
    for line in file:
        line=line.strip()
        A.append(line)

coor = []
for r, row in enumerate(A):
    for c, node in enumerate(row):
        if node == '#' :
            coor.append((r, c))
R,C=len(A),len(A[0])
ER,EC = [], [] # empty rs and cs
IA = [ [A[r][c] for r in range(R)] for c in range(C) ]
for i, r in enumerate(A):
    if any(node != nul for node in r):
        continue
    ER.append(i)
for i, c in enumerate(IA):
    if any(node != nul for node in c):
        continue
    EC.append(i)

def calc(coor,xp=2) -> int:
    res = 0
    for i, node in enumerate(coor):
        r, c = node
        for node_next in coor[:i]:
            next_r, next_c = node_next
            s,e = sorted([r, next_r])
            for node in range(s, e):
                if node in ER:
                    res += xp - 1
            res += e - s
            s,e = sorted([c, next_c])
            for node in range(s, e):
                if node in EC:
                    res += xp - 1
            res += e - s
        #yield res
    return res

r2 = calc(coor,int(1e6))
print('part 1:', calc(coor))
print('part 2:', r2)
# 8871874186836 . 25930568

